packages/bsky/src/vault.ts

```diff
@@ -0,0 +1,57 @@
+import { BsyncClient } from './bsync'
+import { Method } from './proto/bsync_pb'
+
+export const createVaultClient = (bsyncClient: BsyncClient): VaultClient => {
+  return new VaultClient(bsyncClient)
+}
+
+// VaultClient is an abstraction over the BsyncClient, that uses the bsync `PutOperation` RPC
+// to store private data, which can be indexed by the dataplane and queried by the appview.
+export class VaultClient {
+  constructor(private readonly bsyncClient: BsyncClient) {}
+
+  create(input: CreateInput) {
+    return this.putOperation(Method.CREATE, input)
+  }
+
+  update(input: UpdateInput) {
+    return this.putOperation(Method.UPDATE, input)
+  }
+
+  delete(input: DeleteInput) {
+    return this.putOperation(Method.DELETE, { ...input, payload: undefined })
+  }
+
+  private async putOperation(method: Method, input: PutOperationInput) {
+    const { actorDid, namespace, key, payload } = input
+    await this.bsyncClient.putOperation({
+      actorDid,
+      namespace,
+      key,
+      method,
+      payload: payload ? Buffer.from(JSON.stringify(payload)) : undefined,
+    })
```

Comment:
I might recommend this vault abstraction apply lexicon validation based on `namespace` and work with the lexicon data model.  That would probably involve passing a `LexValue` as the `payload`, and using `lexToJson()` or `stringifyLex()` in here.  This would be consistent with typical handling of e.g. record values, xrpc body/input values, etc.

Reply:
You also may want to consider whether the payload explicitly should or shouldn't include `$type`.  In the public repository we do enforce `$type` on records, partly so that if you come across a free-floating record out of context you can determine what it represents and validate it.